### PR TITLE
COL-45 - Implement base asset library REST services

### PR DESCRIPTION
**Pull request by nicolaasmatthijs**
- Bookmarklet token authentication
- Basic create link service
- Basic create link through bookmarklet service
- Basic get asset service
- Basic list course assets service

https://jira.ets.berkeley.edu/jira/browse/COL-45
https://jira.ets.berkeley.edu/jira/browse/COL-47

### CODE REVIEW

**Comment by nicolaasmatthijs** [comment]
Assigning to @simong for review

@johncrossman @paulkerschen : Feel free to have a look at this as well. Let me know if you have any questions or would like a walkthrough.

File: bower.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "collabosphere",
+  "description": "Collabosphere Instructure Canvas LTI Tools",
+  "version": "0.1.0",
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/ets-berkeley-edu/collabosphere.git"
+  },
+  "dependencies": {
+    "angular": "latest",
```
Comment (simong) [inline bower.json]: I assume this stuff will be used for the UI counterpart to this PR?

Can you add some instructions to the README that explain how to pull in these assets through bower?

Comment (simong) [inline bower.json]: Similarly, can you provide an example Apache config? I assume it'll be a straightforward:
 - ProxyPass /api
 - DocumentRoot ~/path/to/collabosphere/public

File: node_modules/col-tests/lib/util.js [hunk 2]
```diff
@@ -18,12 +18,14 @@ var config = require('config');
 var randomstring = require('randomstring');
 var util = require('util');
 
+var AssetsTestUtil = require('col-assets/tests/util');
+var LtiTestUtil = require('col-lti/tests/util');
```
Comment (simong) [inline node_modules/col-tests/lib/util.js]: Doesn't this create a circular reference?

File: node_modules/col-core/lib/api.js [hunk 3]
```diff
@@ -59,21 +59,94 @@ var initializeServer = function(callback) {
...
+    var userId = req.headers['x-collabosphere-user'];
+    var bookmarkletToken = req.headers['x-collabosphere-token'];
+
+    // If the custom headers are not present, move on through the middleware chain
+    if (!userId || !bookmarkletToken) {
+      return next();
+    }
+
+    // Get the user for the provided bookmarklet token
+    UsersAPI.getUserByBookmarkletToken(userId, bookmarkletToken, function(err, user) {
```
Comment (simong) [inline node_modules/col-core/lib/api.js]: We should probably log the error object here (if any)

File: bower.json (hunk 1 above)
Comment (simong) [inline bower.json]: For example:
```
<VirtualHost collabosphere.berkeley:80>
    # ----------------------------------------------------------------------
    # General Host Settings
    # ----------------------------------------------------------------------

    DocumentRoot {PATH_TO_COLLABOSPHERE_REPO}/static
    ServerName collabosphere.berkeley
    ServerAlias collabosphere.berkeley

    # ----------------------------------------------------------------------
    # Log Settings
    # ----------------------------------------------------------------------

    ErrorLog /usr/local/apache2/logs/collabosphere_error.log
    CustomLog /usr/local/apache2/logs/collabosphere_custom.log common

    # ----------------------------------------------------------------------
    # ProxyPass Directives
    # ----------------------------------------------------------------------

    # Proxy API & LTI calls to the application server
    ProxyPass /api http://localhost:2000/api retry=0
    ProxyPass /lti http://localhost:2000/lti retry=0
    ProxyPreserveHost On

</VirtualHost>
```

Comment (nicolaasmatthijs) [inline bower.json]: I'll make this part of the UI counterpart PR

File: node_modules/col-tests/lib/util.js (hunk 2 above)
Comment (nicolaasmatthijs) [inline node_modules/col-tests/lib/util.js]: It does, but for some reason it appears to be working fine. Do you want me to move `getAssetLibraryClient` back to the LTI module?

File: node_modules/col-assets/lib/api.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+ * software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+var Joi = require('joi');
+
+var DB = require('col-core/lib/db');
+var CollabosphereUtil = require('col-core/lib/util');
```
Comment (simong) [inline node_modules/col-assets/lib/api.js]: This is out of order

File: node_modules/col-assets/tests/util.js [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+  assert.ok(asset);
+  assert.ok(asset.id);
+  assert.ok(asset.type);
+  assert(CollabosphereConstants.ASSET.ASSET_TYPES.indexOf(asset.type) !== -1)
+  assert.ok(asset.course_id);
+  assert.ok(asset.user_id);
+  assert.ok(asset.title);
+  assert.ok(asset.created_at);
+  assert.ok(asset.updated_at);
+  assert(typeof asset.likes === 'number');
```
Comment (simong) [inline node_modules/col-assets/tests/util.js]: Can we not do: `assert.ok(_.isFinite(asset.likes));` for these guys?

File: node_modules/col-canvas/lib/api.js [hunk 6]
```diff
@@ -26,8 +28,22 @@ var log = require('col-core/lib/logger')('col-canvas');
  * @param  {Canvas}     callback.canvas       The canvas instance
  */
 var getCanvas = module.exports.getCanvas = function(apiDomain, ltiKey, callback) {
-  // TODO: Validation
+  // Parameter validation
+  var validationSchema = Joi.object().keys({
+    'apiDomain': Joi.string().required(),
+    'ltiKey': Joi.string().required()
```
Comment (simong) [inline node_modules/col-canvas/lib/api.js]: we can probably make this even more strict and require a 32 character string

File: node_modules/col-core/lib/server.js [hunk 7]
```diff
@@ -97,6 +97,24 @@ var setUpServer = module.exports.setUpServer = function() {
     return next();
   });
 
+  // Add CORS headers to support Bookmarklet requests.
+  // Cookies won't be passed in so all cross domain requests will be anonymous
+  app.use(function(req, res, next) {
```
Comment (simong) [inline node_modules/col-core/lib/server.js]: We could do this with Apache

File: node_modules/col-tests/lib/util.js (hunk 2 above)
Comment (simong) [inline node_modules/col-tests/lib/util.js]: It's probably OK, but the time the exported functions get executed we should be in a next tick and each method should be available (I think)

**Comment by simong** [comment]
LGTM, some minor follow-up can be done as part of the next PR

File: node_modules/col-assets/tests/util.js (hunk 5 above)
Comment (nicolaasmatthijs) [inline node_modules/col-assets/tests/util.js]: We sure can. I'll switch to `_.isBoolean` as well

File: node_modules/col-core/lib/server.js (hunk 7 above)
Comment (nicolaasmatthijs) [inline node_modules/col-core/lib/server.js]: Okay, cool, I'll make that part of the PR that introduces Apache and the production build

**Comment by nicolaasmatthijs** [comment]
The follow-up to the review has been incorporated in my COL-50 branch

**Pull request merged**